File: Labs/Lab09/lab9-monte_carlo.py
# Monte Carlo integration - evaluates f(x) at randomly sampled x
#
# by Denes Molnar for PHYS 580, Fall 2020
#

import random
from math import sqrt, pi as PI
import numpy as np


import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt



# function f(x) to integrate
def f(x): return sqrt(4. - x * x)


# read simulation parameters
n = int( input('Number of random x values per trial: ') )
Ntrials = int( input('Number of independent trials: ') )

x1, x2 = 0., 2.   # integration range

# initialize RNG
# - by default, Python uses the Mersenne Twiste, which is good enough for us
#   => nothing to do



# plots 
#

#fontsize = 30   # fontsize, need to be tuned to screen resolution
fontsize = 15    # fontsize, for 1920x1080
linewidth = 1.5
#linewidth = 3
pointarea = 20.
#pointarea = 50.


# one 2D figure
fig, axArray = plt.subplots(1, 1)
(ax0) = axArray

ax0.set_xlim((x1, x2))
#ax0.set_ylim((0, 1))
ax0.set_xlabel("x")
ax0.set_ylabel("f")


# turn on interactive mode, so plt.show() does not block, then create plot window
matplotlib.pyplot.ion()
plt.show()


# plot the integrand vs x
#
n1 = 1000  # number of points to use
dx = (x2 - x1) / n1
xvals = np.array( [ i * dx   for i in range(n1 + 1) ] )
yvals = np.array( [ f(x)     for x in xvals ] )

ax0.plot(xvals, yvals, linewidth = linewidth, color = "blue", label = "f(x)")
plt.pause(0.01) # show curve


# do the Monte Carlo integration
# - cycle colors in visualization
#

colors = ["red", "orange", "black", "green", "cyan", "yellow"]
Ncolors = len(colors)
xx = np.zeros(n)
yy = np.zeros(n)

sum, sum2 = 0., 0.    # sum and sum-squared for integration results obtained in trials
#displayFreq = 1000
displayFreq = 50
for t in range(Ntrials):
   ax0.set_title("Monte Carlo Integration, trial " + str(t))
   # collect sum(f), do the integral
   tot = 0.
   for i in range(n):
      x = x1 + random.random() * (x2 - x1)  # uniform on [x1,x2]
      fx = f(x)
      xx[i], yy[i] = x, fx   # store for plotting
      tot += fx
   result = tot * (x2 - x1) / n
   # show set of points used in every displayFreq-th trial
   #
   # - if things run too slowly, DISABLE this by setting a high displayFreq 
   #
   if (t % displayFreq) == 0:
      colIdx = (t // displayFreq) % Ncolors   # cycle through colors
      ax0.scatter(xx, yy, s = pointarea, color = colors[colIdx])
      plt.pause(0.1)
   # update sums for statistical analysis
   sum += result
   sum2 += result**2

# print average and standard deviation over trials
#

mean = sum / Ntrials
var = sum2 / Ntrials - mean**2
stddev = sqrt(var)
err = stddev / sqrt(Ntrials)

print("n=%d, Ntrials=%d" % (n, Ntrials) )
print("mean=%0.5f, stddev=%0.5f, err=%0.5f" % (mean, stddev, err) )
print("exact=", PI)


# keep plots open until user interaction
input("[ENTER]")


#EOF

File: Labs/Lab09/Lab9_montecarlo.m
% Monte Carlo integration - evaluates f(x) at randomly sampled x
%

clear;
clf;

% integrand f(x), defined as an "anonymous function" in Matlab
%
f = @(x) sqrt(4. - x*x);

% read computational parameters
%
n = input('Number of random x values per trial = ');
Ntrials = input('Number of independent trials = ');

% initialize RNG 
% - Matlab uses the Mersenne Twister algorithm by default, which is good enough for us
%
%s = RandStream('mt19937ar','Seed',0);
%RandStream.setGlobalStream(s);



% plot the integrand function
%
x1 = 0;
x2 = 2;
n1 = 1000;
dx = (x2 - x1) / n1;
x = zeros(n + 1, 1);
y = zeros(n + 1, 1);
for i = 1:(n1 + 1)
   x(i) = x1 + (i - 1) * dx;
   y(i) = f(x(i));
end
plot(x, y, 'b');
hold on;
xlabel('x');
ylabel('y');


% plot the MC integration progress in colors - turn plots off for speed
%
col(1) = 'r';
col(2) = 'm';
col(3) = 'g';
col(4) = 'b';
col(5) = 'y';
sum = 0;
sum2 = 0;
for j = 1:Ntrials
   r = 0;
   title(['Monte Carlo Integration I, trial = ',num2str(j)]);
   for i = 1:n
      t = rand * (x2 - x1) + x1;
      v = f(t);
      xx(i) = t;
      yy(i) = v;
      r = r + v;
   end
   %% COMMENT THIS BLOCK OUT FOR SPEED
   if (j == 1 || mod(j, 51) == 0);  % only plotting every 51st trial
      k = mod(j, 5) + 1;
      scatter(xx, yy, col(k), '.');
      pause(0.1);
   end
   %% UNTIL HERE
   s = r * (x2 - x1) / n;
   sum = sum + s;
   sum2 = sum2 + s^2;
end

% simple statistics in command window. "err" refers to the standard error of the mean
%
mean = sum / Ntrials;
var = sum2 / Ntrials - mean^2;
sd = sqrt(var);
err = sd / sqrt(Ntrials);
X = sprintf(' n = %d, trials = %d, mean = %0.5f, sd = %0.5f, err = %0.5f', ...
            n, Ntrials, mean, sd, err);
disp(X);

% compare to numerical integration by quadrature
% - Matlab has a routine that does just that
%
v = quad(f, 0, 2);
Y = sprintf(' Matlab Quadrature Result = %0.10f', v);
disp(Y);

